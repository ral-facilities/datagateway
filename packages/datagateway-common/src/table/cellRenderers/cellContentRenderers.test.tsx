import React from 'react';
import { shallow } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { Link } from '@mui/material';
import {
  formatBytes,
  datasetLink,
  investigationLink,
  tableLink,
  formatCountOrSize,
  getStudyInfoInvestigation,
} from './cellContentRenderers';

describe('Cell content renderers', () => {
  describe('formatBytes', () => {
    it('converts to bytes correctly', () => {
      expect(formatBytes(10000)).toEqual('10 KB');
    });

    it('handles 0 correctly', () => {
      expect(formatBytes(0)).toEqual('0 B');
    });

    it('handles -1 correctly', () => {
      expect(formatBytes(-1)).toEqual('Loading...');
    });

    it('handles negative numbers correctly', () => {
      expect(formatBytes(-7)).toEqual('Unknown');
    });

    it('handles undefined correctly', () => {
      expect(formatBytes(undefined)).toEqual('Unknown');
    });
  });

  describe('formatCountOrSize', () => {
    it('Returns calculating if query is fetching', () => {
      expect(formatCountOrSize({ isFetching: true })).toEqual('Calculating...');
    });

    it('Returns unknown if query result is undefined', () => {
      expect(formatCountOrSize({ isFetching: false, data: undefined })).toEqual(
        'Unknown'
      );
      expect(formatCountOrSize({ isFetching: false })).toEqual('Unknown');
      expect(formatCountOrSize({})).toEqual('Unknown');
      expect(formatCountOrSize(undefined)).toEqual('Unknown');
    });

    it('Returns data if query is successful', () => {
      expect(
        formatCountOrSize({ isFetching: false, isSuccess: true, data: 1 })
      ).toEqual('1');
      expect(formatCountOrSize({ data: 1, isSuccess: true })).toEqual('1');
    });

    it('Returns data formatted in bytes when byte flag is set', () => {
      expect(
        formatCountOrSize({ isFetching: false, isSuccess: true, data: 1 }, true)
      ).toEqual('1 B');
      expect(formatCountOrSize({ data: 10000, isSuccess: true }, true)).toEqual(
        '10 KB'
      );
    });

    it('Returns data if query is successful and result is zero', () => {
      expect(
        formatCountOrSize({ isFetching: false, isSuccess: true, data: 0 })
      ).toEqual('0');
    });
  });

  describe('getStudyInfoInvestigation', () => {
    it('filters out missing investigations and returns first existing investigation', () => {
      expect(
        getStudyInfoInvestigation({
          id: 1,
          pid: 'doi 1',
          name: 'study 1',
          modTime: '',
          createTime: '',
          studyInvestigations: [
            {
              id: 2,
            },
            {
              id: 3,
              investigation: {
                id: 4,
                title: 'Investigating the properties of the number 4',
                name: 'investigation 4',
                visitId: '1',
              },
            },
          ],
        })?.name
      ).toEqual('investigation 4');
    });

    it('handles undefined properties fine', () => {
      expect(
        getStudyInfoInvestigation({
          id: 1,
          pid: 'doi 1',
          name: 'study 1',
          modTime: '',
          createTime: '',
        })
      ).toBeUndefined();
      expect(
        getStudyInfoInvestigation({
          id: 1,
          pid: 'doi 1',
          name: 'study 1',
          modTime: '',
          createTime: '',
          studyInvestigations: [
            {
              id: 2,
            },
            {
              id: 3,
            },
          ],
        })
      ).toBeUndefined();
    });
  });

  describe('datasetLink', () => {
    it('renders correctly', () => {
      const wrapper = shallow(
        <MemoryRouter>{datasetLink('1', 2, 'test', 'card')}</MemoryRouter>
      );
      expect(wrapper.find(Link)).toMatchSnapshot();
    });
  });

  describe('investigationLink', () => {
    it('renders correctly', () => {
      const wrapper = shallow(
        <MemoryRouter>{investigationLink(1, 'test', 'card')}</MemoryRouter>
      );
      expect(wrapper.find(Link)).toMatchSnapshot();
    });
  });

  describe('tableLink', () => {
    it('renders correctly', () => {
      const wrapper = shallow(
        <MemoryRouter>{tableLink('/test/url', 'test text')}</MemoryRouter>
      );
      expect(wrapper.find(Link)).toMatchSnapshot();
    });
  });
});
