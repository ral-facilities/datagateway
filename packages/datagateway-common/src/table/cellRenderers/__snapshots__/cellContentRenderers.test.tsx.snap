// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell content renderers datasetLink renders correctly 1`] = `
<ForwardRef(Link)
  component={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "propTypes": Object {
        "innerRef": [Function],
        "onClick": [Function],
        "replace": [Function],
        "target": [Function],
        "to": [Function],
      },
      "render": [Function],
    }
  }
  to="/browse/investigation/1/dataset/2/datafile?view=card"
>
  test
</ForwardRef(Link)>
`;

exports[`Cell content renderers investigationLink renders correctly 1`] = `
<ForwardRef(Link)
  component={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "propTypes": Object {
        "innerRef": [Function],
        "onClick": [Function],
        "replace": [Function],
        "target": [Function],
        "to": [Function],
      },
      "render": [Function],
    }
  }
  to="/browse/investigation/1/dataset?view=card"
>
  test
</ForwardRef(Link)>
`;

exports[`Cell content renderers tableLink renders correctly 1`] = `
<ForwardRef(Link)
  component={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "propTypes": Object {
        "innerRef": [Function],
        "onClick": [Function],
        "replace": [Function],
        "target": [Function],
        "to": [Function],
      },
      "render": [Function],
    }
  }
  to="/test/url"
>
  test text
</ForwardRef(Link)>
`;
